sql: collated string column values (phase two of collation support)

eisenstatdavid opened the pull request:
Columns can now have collated string types.

DO NOT MAKE INDEXES OR PRIMARY KEYS with collated strings unless you're
willing to wipe the database after the planned incompatible
serialization changes in phase three of collation support.

Tracking issue for collation: #2473

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/12294)
<!-- Reviewable:end -->

bdarnell commented:
I'd rather not have a beta release with a feature that, if used, will force the database to be wiped on a future upgrade. Can we make it an error to use collated columns in an index instead of silently doing the wrong thing? (unless this PR and its follow-up will both be merged in the same week)

knz commented:
Yeah +1 on what ben says, please implement a check that rejects the unsupported constructs with an error message.

Also:
- please extend the commit message to explain in general terms that you've replaced uses of `ColumnType_Kind` by `ColumnType`, which contains a `ColumnType_Kind` as attribute, because now the methods that work on ColumnTypes also need the other attributes than Kind (especially the Locale) to decide what to do.
- also extend the commit message to explain the assumption you're making that strings with the same collation key compare equal and what are the limits of this assumption.

cc @RaduBerinde for the proposed changes to the distsql protobufs.

---

Reviewed 35 of 35 files at r1.
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/sql/parser/create.go, line 227 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12294#-KYs4xxH68x14B1fhvva:-KYs4xxH68x14B1fhvvb:b-kgr1bh) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d923969a29b9e310ddd2b287902e706ebfa15ec/pkg/sql/parser/create.go#L227)):*
> ```Go
> 				return nil, errors.Wrapf(err, "invalid locale %s", locale)
> 			}
> 			if s, ok := d.Type.(*StringColType); ok {
> ```

use a type `switch` here.

---

*[pkg/sql/sqlbase/table.go, line 489 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12294#-KYs6cJN_yUAYUOLzMWW:-KYs6cJN_yUAYUOLzMWX:bt1e6ke) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d923969a29b9e310ddd2b287902e706ebfa15ec/pkg/sql/sqlbase/table.go#L489)):*
> ```Go
> 		return b, nil
> 	case *parser.DCollatedString:
> 		// TODO(eisen): assume for now that strings collating equal are identical...
> ```

Please expand this TODO to suggest what you're thinking about, by e.g. giving an example of when this is technically false.

---

*[pkg/sql/sqlbase/table.go, line 1046 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12294#-KYs6z06eKY-9xWDg0hl:-KYs6z073l-1wuhim7-J:b-apayz0) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d923969a29b9e310ddd2b287902e706ebfa15ec/pkg/sql/sqlbase/table.go#L1046)):*
> ```Go
> 	default:
> 		if typ, ok := valType.(parser.TCollatedString); ok {
> 			// TODO(eisen): assume for now that strings collating equal are
> ```

See my comment above.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12294#-:-KYs3uOTPKzItIT16m1l:b-kxox9k)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented on pkg/sql/sqlbase/testutils.go:
```diff
@@ -103,25 +105,49 @@ func RandDatum(rng *rand.Rand, typ ColumnType_Kind, null bool) parser.Datum {
+		}
+		// Generate a random Unicode string.
+		var buf bytes.Buffer
+		for i := 0; i < 10; i++ {
```
why 10?

vivekmenezes commented on pkg/sql/sqlbase/testutils.go:
```diff
@@ -103,25 +105,49 @@ func RandDatum(rng *rand.Rand, typ ColumnType_Kind, null bool) parser.Datum {
+				r = rune(rng.Intn(unicode.MaxRune + 1))
+				if !unicode.Is(unicode.C, r) {
+					break
+				}
```
could you use of these http://stackoverflow.com/questions/22892120/how-to-generate-a-random-string-of-a-fixed-length-in-golang

?

eisenstatdavid commented:
The easiest way to implement the no-use-in-indexes-or-primary-keys check and be sure that it's correct is to make the key encode/key decode functions panic on collated strings and then generate null whenever the collated string kind comes up.

If that's not acceptable, is it enough to ensure that there are no primary key or unique constraints involving a collated string typed column, and that there are no indexes that include collated columns?

---

Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12294#-:-KYstZdpVA2i1idh1vG5:bjdp22z)*
<!-- Sent from Reviewable.io -->

knz commented:
Regarding the latter question, let's do the panic() for now and document this amply in the release notes.

eisenstatdavid commented:
Done, PTAL

---

Review status: 27 of 35 files reviewed at latest revision, 5 unresolved discussions.

---

*[pkg/sql/parser/create.go, line 227 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12294#-KYs4xxH68x14B1fhvva:-KYzFUBdvB5G8KL9BE3a:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d923969a29b9e310ddd2b287902e706ebfa15ec/pkg/sql/parser/create.go#L227)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

use a type `switch` here.
</blockquote></details>

Done.

---

*[pkg/sql/sqlbase/table.go, line 489 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12294#-KYs6cJN_yUAYUOLzMWW:-KYzFqzA8bFRa2MjYtuO:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d923969a29b9e310ddd2b287902e706ebfa15ec/pkg/sql/sqlbase/table.go#L489)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Please expand this TODO to suggest what you're thinking about, by e.g. giving an example of when this is technically false.
</blockquote></details>

Done.

---

*[pkg/sql/sqlbase/table.go, line 1046 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12294#-KYs6z06eKY-9xWDg0hl:-KYzFrJV9ISFx91iNHV2:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d923969a29b9e310ddd2b287902e706ebfa15ec/pkg/sql/sqlbase/table.go#L1046)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

See my comment above.
</blockquote></details>

Done.

---

*[pkg/sql/sqlbase/testutils.go, line 114 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12294#-KYomr8vA33GZqLCYFWz-r1-114:-KYzGUQf_9ythpjxEOag:bmyd72v) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d923969a29b9e310ddd2b287902e706ebfa15ec/pkg/sql/sqlbase/testutils.go#L114)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

why 10?
</blockquote></details>

Changed to `rng.Intn(10)`, like the others.

---

*[pkg/sql/sqlbase/testutils.go, line 120 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12294#-KYomr8vA33GZqLCYFWz-r1-120:-KYzG7XkW_TPOCFxkEyN:byjriqs) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d923969a29b9e310ddd2b287902e706ebfa15ec/pkg/sql/sqlbase/testutils.go#L120)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

could you use of these http://stackoverflow.com/questions/22892120/how-to-generate-a-random-string-of-a-fixed-length-in-golang

?
</blockquote></details>

The linked answers aren't helpful because they don't generate most code points and speed doesn't really matter for us.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12294#-:-KYzEpp0WYS8NK20_wcB:b-oq3912)*
<!-- Sent from Reviewable.io -->

knz commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 9 of 9 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12294#-:-KZ1TsOrPQldm9451zte:bnfp4nl)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/sql/sqlbase/testutils.go, line 114 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12294#-KYomr8vA33GZqLCYFWz-r1-114:-KZ23ZldeCb8J8JX2QBg:b-75ps9l) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d923969a29b9e310ddd2b287902e706ebfa15ec/pkg/sql/sqlbase/testutils.go#L114)):*
<details><summary><i>Previously, eisenstatdavid (David Eisenstat) wrote…</i></summary><blockquote>

Changed to `rng.Intn(10)`, like the others.
</blockquote></details>

I believe that's going to be evaluated on each iteration

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12294#-:-KZ23i5f_n6z-lvq4mBP:bnfp4nl)*
<!-- Sent from Reviewable.io -->

Commit:
sql: collated string column values (phase two of collation support)

Columns can now have collated string types. DO NOT MAKE INDEXES OR
PRIMARY KEYS with collated strings. CRDB will panic until phase three of
collation support is completed.

Many uses of `ColumnType_Kind` were converted to `ColumnType`, so that
the locale of the COLLATEDSTRING kind could be preserved.

Tracking issue for collation: #2473

eisenstatdavid commented:
Review status: 31 of 36 files reviewed at latest revision, 1 unresolved discussion.

---

*[pkg/sql/sqlbase/testutils.go, line 114 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12294#-KYomr8vA33GZqLCYFWz-r1-114:-KZ27jU5VUcXghxuJBbH:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d923969a29b9e310ddd2b287902e706ebfa15ec/pkg/sql/sqlbase/testutils.go#L114)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

I believe that's going to be evaluated on each iteration
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12294)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
*[pkg/sql/sqlbase/structured.go, line 1665 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/12294#-KZ2heeyt3ttdP3VY_9P:-KZ2heeyt3ttdP3VY_9Q:bt3fzpq) ([raw file](https://github.com/cockroachdb/cockroach/blob/001bfa02ae2963435cfe485b0045e694f7eeebeb/pkg/sql/sqlbase/structured.go#L1665)):*
> ```Go
> 			ctyp.Locale = &t.Locale
> 		}
> 		panic(fmt.Sprintf("unsupported result type: %s", ptyp))
> ```

are we missing an `else {}` here?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12294)*
<!-- Sent from Reviewable.io -->
